lib/inventory/controller.go

```diff
@@ -1087,19 +1128,41 @@ func (c *Controller) handleDatabaseServerHB(handle *upstreamHandle, databaseServ
 		if c.dbHBVariableDuration != nil {
 			c.dbHBVariableDuration.Inc()
 		}
-		handle.databaseServers[dbKey] = &heartBeatInfo[*types.DatabaseServerV3]{}
+		hbInfo := &heartBeatInfo[*types.DatabaseServerV3]{}
+		if handle.Hello().GetCapabilities().GetDatabaseHeartbeatGracefulStop() {
+			// graceful stop for individual heartbeats is necessary to enforce
+			// rate limiting without dropping upserts.
+			// otherwise a client may unregister a single server and delete its
+			// backend heartbeat directly, but a rate limit delayed upsert would
+			// reverse the deletion and keepalive the server indefinitely
+			hbInfo.heartbeatLimiter = c.newServerHBRateLimiter()
+		}
+		handle.databaseServers[dbKey] = hbInfo
 		handle.dbKeepAliveDelay.Add(dbKey)
 	}
 
 	now := time.Now()
+	srv := handle.databaseServers[dbKey]
+	if delay := srv.delayUpsertFrom(now); delay > 0 {
+		c.testEvent(dbUpsertRateLimited)
+		slog.DebugContext(c.closeContext, "Database heartbeat rate limit exceeded",
```

Comment:
This will potentially delay _every_ change to the resource, even changes that must go in immediately for cluster connectivity to work (such as updating the list of proxy IDs).